batch-processing/helpers.go

```diff
@@ -0,0 +1,202 @@
+package main
+
+import (
+	"bytes"
+	"encoding/csv"
+	"fmt"
+
+	"image"
+	_ "image/gif"
+	_ "image/jpeg"
+	_ "image/png"
+	"io"
+	"log"
+	"math/rand"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/ssgreg/bottleneck"
+	"golang.org/x/exp/slices"
+	"gopkg.in/gographics/imagick.v2/imagick"
+)
+
+type ProcessImage struct {
+	url    string
+	input  string
+	output string
+	err    error
+	suffix string
+}
+
+type Channels struct {
+	urlsChan            chan string
+	processingErrorChan chan ProcessImage
+	inputPathsChan      chan ProcessImage
+	outputPathsChan     chan ProcessImage
+}
+
+type ConvertImageCommand func(args []string) (*imagick.ImageCommandResult, error)
+
+type Converter struct {
+	cmd ConvertImageCommand
+}
+
+func genFilepath(out, suffix string) string {
+	if out != "" {
+		return fmt.Sprintf("%d-%d-%s.%s", time.Now().UnixMilli(), rand.Int(), out, suffix)
+	}
+	return fmt.Sprintf("%d-%d.%s", time.Now().UnixMilli(), rand.Int(), suffix)
+}
+
+func (c *Converter) Grayscale(inputFilepath string, outputFilepath string) error {
+	// Convert the image to grayscale using imagemagick
+	// We are directly calling the convert command
+	_, err := c.cmd([]string{
+		"convert", inputFilepath, "-set", "colorspace", "Gray", outputFilepath,
+	})
+	return err
+}
+
+func ReadCsvFile(filename, headerTitle string) ([]string, error) {
+	bc.TimeSlice(bottleneck.Index0)
+	// Open CSV file to read url links
+	fileContent, err := os.Open(filename)
+	if err != nil {
+		return []string{}, fmt.Errorf("couldn't open input file, %v", err)
+	}
+	defer fileContent.Close()
+
+	//read from this file
+	records, err := csv.NewReader(fileContent).ReadAll()
+	if err != nil {
+		return []string{}, fmt.Errorf("couldn't read input file, %v", err)
+	}
+
+	//check if there is any content
+	if len(records) == 0 {
+		return []string{}, fmt.Errorf("empty csv")
+	}
+	header := records[0]
+	// find index of needed header if none return error
+
+	//check if header and title info is what we expect
+	// assuming we have multiple columns in the file, we need to find url
+	var urls []string
+	if header[0] == headerTitle {
+		fmt.Println(records)
+		for n := 1; n < len(records); n++ {
+			urls = append(urls, records[n][0])
+			fmt.Println("read a record")
+		}
+	} else {
+		return []string{}, fmt.Errorf("no matching header found, expected %s, got %v", headerTitle, header)
+	}
```

Comment:
One common pattern, particularly in Go, is to early-return on errors to avoid indentation, and then write the happy-path less indented; so you'd write:
```suggestion
	if header[0] != headerTitle {
		return []string{}, fmt.Errorf("no matching header found, expected %s, got %v", headerTitle, header)
	}
	fmt.Println(records)
	for n := 1; n < len(records); n++ {
		urls = append(urls, records[n][0])
		fmt.Println("read a record")
	}
```

This avoids your code moving more and more to the right as it gets indented, and makes it easier to focus on what you _expect_ to be happening in the code.